src/shared/completion/use-set-form-completion-mutation.js

```diff
@@ -0,0 +1,108 @@
+import { useAlert, useDataEngine } from '@dhis2/app-runtime'
+import i18n from '@dhis2/d2-i18n'
+import { useQueryClient, useMutation } from '@tanstack/react-query'
+import { useApiAttributeParams } from '../use-api-attribute-params.js'
+import {
+    useDataSetId,
+    useOrgUnitId,
+    usePeriodId,
+} from '../use-context-selection/index.js'
+import { useDataValueSetQueryKey } from '../use-data-value-set/index.js'
+
+const MUTATION_SET_FORM_COMPLETION = {
+    resource: 'dataEntry/dataSetCompletion',
+    type: 'create',
+    data: ({
+        dataSetId,
+        periodId,
+        orgUnitId,
+        categoryComboId,
+        categoryOptionIds,
+        completed,
+    }) => ({
+        dataSet: dataSetId,
+        period: periodId,
+        orgUnit: orgUnitId,
+        attribute: { combo: categoryComboId, options: categoryOptionIds },
+        completed,
+    }),
+}
+
+export default function useSetFormCompletionMutation() {
+    const queryClient = useQueryClient()
+    const engine = useDataEngine()
+    const [dataSetId] = useDataSetId()
+    const [periodId] = usePeriodId()
+    const [orgUnitId] = useOrgUnitId()
+    const {
+        attributeCombo: categoryComboId,
+        attributeOptions: categoryOptionIds,
+    } = useApiAttributeParams()
+
+    const { show: showErrorAlert } = useAlert((message) => message, {
+        critical: true,
+    })
+
+    const mutationFn = ({ completed }) => {
+        const variables = {
+            dataSetId,
+            periodId,
+            orgUnitId,
+            categoryComboId,
+            categoryOptionIds,
+            completed,
+        }
+
+        return engine.mutate(MUTATION_SET_FORM_COMPLETION, { variables })
+    }
+
+    const dataValueSetQueryKey = useDataValueSetQueryKey()
```

Comment:
Please take a look how this is done here: https://github.com/dhis2/data-entry-app/blob/development/src/data-workspace/data-value-mutations/mutation-functions.js#L31

We should not need to get the context-params manually again here. Please get them from the `mutationKey`. Also a `mutation-key-factory.js` would be great, to follow the same pattern as `dataValueMutation`. But might not be needed for now.
By getting the parameters from the `mutationKey`, we are guaranteed that the queryKey contains everything that is needed for the mutation, which can prevent subtle bugs. Think of the `mutationKey` as a `dependency`-array to the mutation. 
If you don't think a separate mutationKey is needed here, still try to get the params from the `useDataValueSetQueryKey`. 

```js

//outside hook

function mutateFn({ completed} ) {
        // get params for the mutation from the `mutationKey`
       const { mutationKey } = this
       const { params } = mutationKey[1]
       const variables = {
            dataSetId: params.ds,
            periodId: params.pe,
            orgUnitId: params.ou
            categoryComboId: params.cc
            categoryOptionIds: params.cp
       }
      const {
          ds: dataSetId,
          pe: periodId,
          ou: orgUnitId,
          cc: categoryComboId,
          cp: categoryOptionIds,
       } = mutationKey[1]

        return engine.mutate(MUTATION_SET_FORM_COMPLETION, { variables })
    }

// in hook
   // could create a new hook that returns in non-abbreviated format
   // instead of having to translate this in mutateFn above
const dataValueSetQueryKey = useDataValueSetQueryKey()

return useMutation(mutationFn, {
      retry: 0,
      mutationKey: dataValueSetQueryKey,
      onMutate: ...
```

Is this functionality available offline? If so, we should extract the `mutationFn` (like above) to a separate function and setup a `defaultMutationFn` in `use-query-client`. Similar to how it's done for `dataValues`.

Reply:
This has been addressed by both @Birkbjo as well as @kabaros